InstanceLootPlugin/DropRateManager.cs

```diff
@@ -0,0 +1,77 @@
+using System;
+using BepInEx.Configuration;
+
+namespace InstanceLootPlugin;
+
+public class DropRateManager
+{
+    private ConfigEntry<bool> EnableBadLuckProtection;
+    private ConfigEntry<bool> EnableSwarmsScaling;
+    private ConfigEntry<float> DropChanceMultiplier;
+    private ConfigEntry<float> BaseDropChance;
+    private ConfigEntry<float> MinimumDropChance;
+
+    public DropRateManager(
+        ConfigEntry<bool> enableBadLuckProtection,
+        ConfigEntry<bool> enableSwarmsScaling,
+        ConfigEntry<float> dropChanceMultiplier,
+        ConfigEntry<float> baseDropChance,
+        ConfigEntry<float> minimumDropChance)
+    {
+        EnableBadLuckProtection = enableBadLuckProtection;
+        EnableSwarmsScaling = enableSwarmsScaling;
+        DropChanceMultiplier = dropChanceMultiplier;
+        BaseDropChance = baseDropChance;
+        MinimumDropChance = minimumDropChance;
+    }
+
+    public float GetPlayerAwareBaseDropChance(int playerCount, float dropChance)
+    {
+        float playerBasedDropRateMultiplier = 1f / playerCount;
+        float multiplier = playerBasedDropRateMultiplier * DropChanceMultiplier.Value;
+        float newDropChance = Math.Max(BaseDropChance.Value * multiplier, MinimumDropChance.Value);
+
+        if (!dropChance.Equals(newDropChance))
+        {
+            dropChance = newDropChance;
+            Log.Debug(
+                $"Using dropRate of {dropChance}% (base={BaseDropChance.Value}, multiplier={DropChanceMultiplier.Value}, playerCount={playerCount}, minimumDropChance={MinimumDropChance.Value}");
+        }
+
+        return dropChance;
+    }
+
+    public float ComputeDropChance(bool isSwarmEnabled, float nativeDropChance, float nativeBaseDropChance, float dropChance)
+    {
+        // Approximately balance Swarms Artifact since they didn't do a good job of it.
+        float swarmModifiedNativeChance = !EnableSwarmsScaling.Value || isSwarmEnabled || nativeDropChance == 0
+            ? nativeDropChance
+            : (nativeDropChance - nativeBaseDropChance) * 0.6f + nativeBaseDropChance;
+
+        // Apply the multipliers used by the game to respect the many drop rate factors.
+        float baseModifier = nativeBaseDropChance / 5f;
+        float enemyModifier = swarmModifiedNativeChance / nativeBaseDropChance;
+
+        return dropChance * baseModifier * enemyModifier;
+    }
+
+    public bool NeedsBadLuckProtection(float actualDropChance, float time)
+    {
+        if (EnableBadLuckProtection.Value && DropRateTracker.cumulativeDropChance > 100 && actualDropChance > 0)
```

Comment:
I think you are right in principle, but with the current implementation it wouldn't make a difference because the boosted drop rate is calculated as `dropRate + (cumulativeDropChance - 100)`. I'll make the change; I wouldn't be shocked if bad luck protection was tweaked in the future.

Reply:
Resolved in latest push